Improvements to discriminated union docs (#8042)
pydantic/pydantic: docs/concepts/unions.md

    if isinstance(v, int):
        return 'int'
    if isinstance(v, (dict, BaseModel)):
        return 'model'
    else:
        # return None if the discriminator value isn't found
        return None

class SpecialValue(BaseModel):
    value: int

class DiscriminatedModel(BaseModel):
    value: Annotated[
        Union[
            Annotated[int, Tag('int')],
            Annotated['SpecialValue', Tag('model')],
        ],
        Discriminator(model_x_discriminator),
    ]

model_data = {'value': {'value': 1}}
m = DiscriminatedModel.model_validate(model_data)
print(m)
#> value=SpecialValue(value=1)

int_data = {'value': 123}
m = DiscriminatedModel.model_validate(int_data)
print(m)
#> value=123

try:
    DiscriminatedModel.model_validate({'value': 'not an int or a model'})
except ValidationError as e:
    print(e)  # (1)!
    """
    1 validation error for DiscriminatedModel
    value
      Unable to extract tag using discriminator model_x_discriminator() [type=union_tag_not_found, input_value='not an int or a model', input_type=str]
    """
```

1. Notice the callable discriminator function returns `None` if a discriminator value is not found.

>    When `None` is returned, this error is raised.

Reviewer: ```suggestion
   When `None` is returned, this `union_tag_not_found` error is raised.
```

data = {'x': {'x': {'x': 'a'}}}
m = DiscriminatedModel.model_validate(data)

> print(m)

Author: I believe it's most reader friendly to use print than assert, can we check were else we're using `assert`.

Reviewer 2: Agreed. I'll change the rest of the `assert` statements to `print` statements.

#> x=DiscriminatedModel(x=DiscriminatedModel(x='a'))

> try:

Author: this example, together with an example or the errors raised without a custom error need to be in a separate example.

## Union Validation Errors

> When `Union` validation fails, error messages can be quite verbose, especially when dealing with recursive models.
> Discriminated unions help to simplify error messages in this case.

Reviewer: ```suggestion
When `Union` validation fails, error messages can be quite verbose, as they will produce validation errors for
each case in the union.
This is especially noticeable when dealing with recursive models, where reasons may be generated at each level of
recursion.
Discriminated unions help to simplify error messages in this case, as validation errors are only produced for
the case with a matching discriminator value.
```

except ValidationError as e:
    print(e)
    """
    4 validation errors for Model
    x.str
      Input should be a valid string [type=string_type, input_value={'x': {'x': 1}}, input_type=dict]
    x.Model.x.str
      Input should be a valid string [type=string_type, input_value={'x': 1}, input_type=dict]
    x.Model.x.Model.x.str
      Input should be a valid string [type=string_type, input_value=1, input_type=int]
    x.Model.x.Model.x.Model
      Input should be a valid dictionary or instance of Model [type=model_type, input_value=1, input_type=int]
    """

try:
    Model.model_validate({'x': {'x': {'x': {}}}})
except ValidationError as e:
    print(e)
    """
    4 validation errors for Model
    x.str
      Input should be a valid string [type=string_type, input_value={'x': {'x': {}}}, input_type=dict]
    x.Model.x.str
      Input should be a valid string [type=string_type, input_value={'x': {}}, input_type=dict]
    x.Model.x.Model.x.str
      Input should be a valid string [type=string_type, input_value={}, input_type=dict]
    x.Model.x.Model.x.Model.x
      Field required [type=missing, input_value={}, input_type=dict]
    """

> # Errors are much more simple with a discriminated union:

Reviewer: ```suggestion
# Errors are much simpler with a discriminated union:
```
I know it's not currently affected by this PR but 🤷‍♂️
